execution/executor-benchmark/src/pipeline.rs

```diff
@@ -76,46 +88,115 @@ where
             (None, None)
         };
 
-        let exe_thread = std::thread::Builder::new()
-            .name("txn_executor".to_string())
-            .spawn(move || {
-                start_execution_rx.map(|rx| rx.recv());
-
-                let mut exe = TransactionExecutor::new(
-                    executor_1,
-                    parent_block_id,
-                    version,
-                    Some(commit_sender),
-                    config.allow_discards,
-                    config.allow_aborts,
-                );
-                let start_time = Instant::now();
-                let mut executed = 0;
-                let start_gas = TXN_GAS_USAGE.get_sample_sum();
-                while let Ok(transactions) = block_receiver.recv() {
-                    info!("Received block of size {:?} to execute", transactions.len());
-                    executed += transactions.len();
-                    exe.execute_block(transactions);
-                    info!("Finished executing block");
-                }
+        let mut join_handles = vec![];
 
-                let delta_gas = TXN_GAS_USAGE.get_sample_sum() - start_gas;
-
-                let elapsed = start_time.elapsed().as_secs_f32();
-                info!(
-                    "Overall execution TPS: {} txn/s (over {} txns)",
-                    executed as f32 / elapsed,
-                    executed
-                );
-                info!(
-                    "Overall execution GPS: {} gas/s (over {} txns)",
-                    delta_gas as f32 / elapsed,
-                    executed
-                );
-
-                start_commit_tx.map(|tx| tx.send(()));
-            })
-            .expect("Failed to spawn transaction executor thread.");
+        let mut partitioning_stage = BlockPartitioningStage::new(num_partitioner_shards);
+
+        let mut exe = TransactionExecutor::new(
+            executor_1,
+            parent_block_id,
+            version,
+            Some(commit_sender),
+            config.allow_discards,
+            config.allow_aborts,
+        );
+
+        if config.async_partitioning {
```

Comment:
There is a lot of code duplication here, can you look into refactoring the common code into functions so that we can reduce the duplication?

Reply:
That's why I implemented it [this way](https://github.com/aptos-labs/aptos-core/blob/56ce342d0906a1a5d7875ed10bbc89cbb9c75fed/execution/executor-benchmark/src/pipeline.rs#L101-L178) at the beginning...

Reply:
Hmm I actually don't know a good way to dedup the current code without killing the readability. I will leave it as-is for now and merge. Let's discuss further and make a separate change if needed.

Reply:
IIUC, you should be able to avoid duplication by refactoring the partitioning and execution inside a thread to separate functions. 

```
fn partitioning_block () {

while let Ok(txns) = raw_block_receiver.recv() {
                        let exe_block_msg = partitioning_stage.process(txns);
                        executable_block_sender.send(exe_block_msg).unwrap();
                        
  }

fn execute_block() {

start_execution_rx.map(|rx| rx.recv());
                    let start_time = Instant::now();
                    let mut executed = 0;
                    let start_gas = TXN_GAS_USAGE.get_sample_sum();
                    while let Ok(msg) = executable_block_receiver.recv() {
                        let ExecuteBlockMessage {
                            current_block_start_time,
                            partition_time,
                            block,
                        } = msg;
                        let block_size = block.transactions.num_transactions();
                        info!("Received block of size {:?} to execute", block_size);
                        executed += block_size;
                        exe.execute_block(current_block_start_time, partition_time, block);
                        info!("Finished executing block");
                        
    }
    
    ```

Reply:
I don't think these 2 functions can't be used in sync-partition mode: they each are looping.

Reply:
@sitalkedia  this is the what I came up following your idea.
https://github.com/aptos-labs/aptos-core/commit/2dcd59f0dfaa92f062818f9b5cd8bb8c5adc01d4

Reply:
LGTM - thanks for taking care of it.